File: leetcode_75/Heap_Priority_Queue/python/2336.py
import heapq


class SmallestInfiniteSet:
    def __init__(self):
        self.heap = []
        heapq.heapify(self.heap)
        self.added_back = set()
        self.next_smallest = 1

    def popSmallest(self) -> int:
        if self.heap:
            smallest = heapq.heappop(self.heap)
            self.added_back.remove(smallest)
            return smallest

        smallest = self.next_smallest
        self.next_smallest += 1
        return smallest

    def addBack(self, num: int) -> None:
        if num < self.next_smallest and num not in self.added_back:
            heapq.heappush(self.heap, num)
            self.added_back.add(num)


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)

File: leetcode_75/Heap_Priority_Queue/cpp/2336.cpp
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

class SmallestInfiniteSet
{

    priority_queue<int, vector<int>, greater<int>> heap;
    unordered_set<int> set_heap;
    int next_smallest = 1;

public:
    SmallestInfiniteSet()
    {
    }

    int popSmallest()
    {
        int smallest;

        if (!this->heap.empty())
        {
            smallest = this->heap.top();
            this->heap.pop();
            this->set_heap.erase(smallest);

            return smallest;
        }

        smallest = this->next_smallest;
        this->next_smallest++;

        return smallest;
    }

    void addBack(int num)
    {
        if (num < this->next_smallest && this->set_heap.find(num) == this->set_heap.end())
        {
            this->heap.push(num);
            this->set_heap.insert(num);
        }
    }
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */